File: programs/spectral/Spectral.java
public class Spectral {
    static double evalA(int i, int j) {
        int ij = i + j;
        return 1.0 / (double) ((ij * (ij + 1) / 2) + i + 1);
    }

    static void timesAv(double[] v, double[] out, int n) {
        for (int i = 0; i < n; i++) {
            double s = 0.0;
            for (int j = 0; j < n; j++) s += evalA(i, j) * v[j];
            out[i] = s;
        }
    }

    static void timesA(double[] v, double[] out, int n) {
        for (int i = 0; i < n; i++) {
            double s = 0.0;
            for (int j = 0; j < n; j++) s += evalA(j, i) * v[j];
            out[i] = s;
        }
    }

    public static void main(String[] args) {
        int n = 300;
        double[] u = new double[n], v = new double[n], w = new double[n];
        for (int i = 0; i < n; i++) u[i] = 1.0;
        for (int it = 0; it < 10; it++) {
            timesAv(u, w, n);
            timesA(w, v, n);
            timesAv(v, w, n);
            timesA(w, u, n);
        }
        double vbv = 0.0, vv = 0.0;
        for (int k = 0; k < n; k++) { vbv += u[k] * v[k]; vv += v[k] * v[k]; }
        System.out.println("spectral checksum = " + (long) (vbv / vv * 1000000.0));
    }
}

File: programs/spectral/spectral.rs
fn eval_a(i: i64, j: i64) -> f64 {
    let ij = i + j;
    1.0 / (((ij * (ij + 1) / 2) + i + 1) as f64)
}

fn times_av(v: &[f64], out: &mut [f64], n: usize) {
    let mut i = 0;
    while i < n {
        let mut s = 0.0;
        let mut j = 0;
        while j < n {
            s += eval_a(i as i64, j as i64) * v[j];
            j += 1;
        }
        out[i] = s;
        i += 1;
    }
}

fn times_a(v: &[f64], out: &mut [f64], n: usize) {
    let mut i = 0;
    while i < n {
        let mut s = 0.0;
        let mut j = 0;
        while j < n {
            s += eval_a(j as i64, i as i64) * v[j];
            j += 1;
        }
        out[i] = s;
        i += 1;
    }
}

fn main() {
    let n: usize = 300;
    let mut u = vec![1.0; n];
    let mut v = vec![0.0; n];
    let mut w = vec![0.0; n];
    let mut it = 0;
    while it < 10 {
        times_av(&u, &mut w, n);
        times_a(&w, &mut v, n);
        times_av(&v, &mut w, n);
        times_a(&w, &mut u, n);
        it += 1;
    }
    let mut vbv = 0.0;
    let mut vv = 0.0;
    let mut k = 0;
    while k < n {
        vbv += u[k] * v[k];
        vv += v[k] * v[k];
        k += 1;
    }
    println!("spectral checksum = {}", (vbv / vv * 1000000.0) as i64);
}
